File: Java/102.binaryTreeLevelOrderTraversal.java
// BFS
// class Solution {
//   public List<List<Integer>> levelOrder(TreeNode root) {
//     List<List<Integer>> res = new ArrayList<>();
//     Queue<TreeNode> q = new LinkedList<>();
//     q.offer(root);
//     while (!q.isEmpty()) {
//       int size = q.size();
//       List<Integer> level = new LinkedList<>();
//       for (int i = 0; i < size; ++i) {
//         TreeNode cur = q.peek();
//         q.poll();
//         if (cur == null) {
//           continue;
//         }
//         level.add(cur.val);
//         q.offer(cur.left);
//         q.offer(cur.right);
//       }
//       if (!level.isEmpty()) {
//         res.add(level);
//       }
//     }
//     return res;
//   }
// }

// DFS
class Solution {
  public List<List<Integer>> levelOrder(TreeNode root) {
    List<List<Integer>> res = new ArrayList<>();
    if (root != null) {
      dfs(res, root, 0);
    }
    return res;
  }
  public void dfs(List<List<Integer>> res, TreeNode node, int level) {
    if (res.size() - 1 < level) {
      res.add(new ArrayList<>());
    }
    res.get(level).add(node.val);
    if (node.left != null) {
      dfs(res, node.left, level + 1);
    }
    if (node.right != null) {
      dfs(res, node.right, level + 1);
    }
  }
}

File: Python/102.binaryTreeLevelOrderTraversal.py
# BFS
# class Solution:
#     def levelOrder(self, root: TreeNode) -> List[List[int]]:
#       queue = collections.deque()
#       queue.append(root)
#       res = []
#       while queue:
#         size = len(queue)
#         level = []
#         for _ in range(size):
#           cur = queue.popleft()
#           if not cur:
#             continue
#           level.append(cur.val)
#           queue.append(cur.left)
#           queue.append(cur.right)
#         if level:
#           res.append(level)
#       return res

# DFS
class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
      res = []
      self.level(root, 0, res)
      return res
    def level(self, root, level, res):
      if not root:  return
      if len(res) == level: res.append([])
      res[level].append(root.val)
      if root.left:
        self.level(root.left, level + 1, res)
      if root.right:
        self.level(root.right, level + 1, res)